### ISSUE TITLE
header-layout: problems w/ tab order

### ISSUE THREAD

**Issue by robrez** [PolymerLabs/app-layout#109]
When using app-header-layout, if any element in the main content area (`#contentContainer`) has a tabindex, it will receive focus before anything inside the app-header.

You can see this occur via `app-layout/templates/shrine/index.html`

Here is some example code (slightly modified `app-layout/demo/demo2.html`) --
```
    <app-header-layout>

      <app-header fixed effects="waterfall">
        <app-toolbar>
          <paper-icon-button icon="menu" drawer-toggle></paper-icon-button>
          <div spacer title>My Drive</div>
          <button id="one">Focus me first?</button>
        </app-toolbar>
      </app-header>

      <div content>
        <sample-content size="100"></sample-content>
        <button id="two">Wrong focus order</button>
      </div>

    </app-header-layout>
```

I've only been able to workaround this by moving `#contentContainer` such that is the last element in `app-header-layout` and assigning explicit z-indexing in the css rules to ensure contentContainer always scrolls below the header.

_edit_ explicit z-indexing screws up some things throughout the app. one example is paper-checkbox's ink focus behavior showing up behind the content

**Comment by keanulee** [comment]
@robrez We reorder the elements such that `app-header` comes after the content so that it renders above the content without using `z-index` (which is dangerous for a library to use since we don't know what other `z-index`s you use in your app). If you're okay with setting your own `z-index`, I would suggest not using `app-header-layout` and just creating your own content container with some `padding-top`.

**Comment by robrez** [comment]
@keanulee I definitely would prefer to not set z-index... it seems like I'd have to choose between two things I'd rather not do:
- setting z-index
- setting tabindex

I think I'll do neither and see if I can manage to convince product managers and WCAG compliance folks that content taking tabindex focus before toolbar content is a _feature_. "You really want the user to focus on the content, not the navigation!"

**Comment by robdodson** [comment]
I think this should be promoted to a p0 issue. Any usage of app-header-layout effectively breaks the accessibility for a page

**Comment by robdodson** [comment]
@robrez 

> I think I'll do neither and see if I can manage to convince product managers and WCAG compliance folks that content taking tabindex focus before toolbar content is a feature. "You really want the user to focus on the content, not the navigation!"

This doesn't sound like a good solution to me as it violates WCAG 2.4.3. You want your website to have a logical focus order. To help users jump to content you can use something like a [skip link](http://webaim.org/techniques/skipnav/). Many major websites (github included) use this technique. I think (gross as it may seem) app-header-layout needs to supply a z-index. I know the team is also pushing for better primitives in the platform to work around this but in the near term I'm not sure if there's a better solution than z-index.

**Comment by robrez** [comment]
Thanks for the input

I did end up using z-indexing. It ended up being the header's `:after` rule (specifically that it uses absolute positioning) for waterfall shadow, not z-indexing that was "messing up" (by hiding it) paper-ink-focus behavior... I took a slightly different approach for shadow as well (normal positioning, display block)

After combining these two small tweaks, the product is working well.

We have to be compliant w/ government mandates (which require wcag AA or better)... I feel confident that we have now.

I'm really excited about having customers beta our 100% polymer 'patient portal' application and getting it through "meaningful use" certification

**Comment by robdodson** [comment]
> We have to be compliant w/ government mandates (which require wcag AA or better)... I feel confident that we have now.

> I'm really excited about having customers beta our 100% polymer 'patient portal' application and getting it through "meaningful use" certification

That's great to hear! So I understand a bit better, did you end up making your own container and just using app-header with it?

**Comment by robrez** [comment]
I did patch app-layout with success, but I didn't want to make use of a patched version if the polymer team didn't like the general approach...

So I used:

- my own `position: fixed` header
- `iron-scroll-target-behavior`, to help toggle a "condenses" class on the header
- `app-layout-drawer` for "small" screen navigation

I would normally have furthered the discussion w/ the poly team but was kind of in that phase of the project where upper mgt is like "soooooo is it done yet? Can it be done by...."

**Comment by keanulee** [comment]
@robdodson In short `app-header` must be placed over the content. `app-header-layout` tried to do this without using z-index (because z-index numbers are global and a library can't make assumptions on other z-indicies you use on the page), so it reorders the DOM so that app-header comes after which breaks tab order.

So basically there's not much we can do other than advising users to not use `app-*-layout` and setting their own z-indicies. This is unfortunate, but it's a limitation of the platform today.

WDYT @blasten @frankiefu ?

**Comment by frankiefu** [comment]
Seems like lots of people need to workaround this issue by **not** using `app-header-layout`, I'm inclined that we should consider changing `app-header-layout` to use `z-index` to make it work for accessibility.  This is unfortunate but also we should provide something that users can use in their app without making too many custom changes.

Also one of the goal of having `*-layout` element is that we can have different layouts.  In this case, maybe we can create another `*-layout` element which has the correct DOM orders along with using z-index to layout the elements.

**Comment by robrez** [comment]
I think the app-layout scaffold is awesome, no doubt it will be widely used.

What about an opt-in bring-your-own-z-index model of some sort?

**Comment by robdodson** [comment]
+1 to everything @frankiefu said.

I'm not sure if paper-styles already does this but we may want to expose a set of z-indexs that everyone abides by in their app. Check out this medium post https://medium.com/@fat/mediums-css-is-actually-pretty-fucking-good-b8e2a6c78b06#.o6bfiis23 and look for the heading "The 2nd Project: Scales" (no idea how to deep link in a medium post :P)

**Comment by frankiefu** [comment]
Fixed in https://github.com/PolymerLabs/app-layout/pull/146
